File: exante/src/main/java/exante/Task2.java
package exante;

import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;

/**
 * Максимальное кол-во повторяющихся квадратных корней
 */
public class Task2 {

    static int UPPER_LIMIT = 1_000_000_000;

    static final SortedMap<Integer, SortedSet<Long>> squareMap = new TreeMap<>((a,b) -> b-a);

    @Test
    public void test() {
        assertEquals(4, solution(2, UPPER_LIMIT));
        //assertEquals(3, solution(6000, 7000));
        //assertEquals(1, solution1(3, 5));
        assertEquals(2, solution1(16, 16));
    }

    public static void main(String[] args) {
        assertEquals(4, solution(2, UPPER_LIMIT));
        //assertEquals(3, solution(6000, 7000));
        //assertEquals(1, solution1(3, 5));
        assertEquals(2, solution1(16, 16));
    }

    static void clearAndFillMap() {
        squareMap.clear();
        // fill 0
        fill0();
        int squareRootNumber = 1;
        while (true) {
            SortedSet<Long> previousSquares = squareMap.get(squareRootNumber-1);
            long first = previousSquares.first();
            if (first*first > UPPER_LIMIT) {
                break;
            }
            SortedSet<Long> squares = new TreeSet<>();
            for (long s : previousSquares) {
                long ss = s*s;
                if (ss > UPPER_LIMIT) {
                    break;
                }
                squares.add(ss);
            }
            squareMap.put(squareRootNumber, squares);
            squareRootNumber++;
        }
    }

    private static final void fill0() {
        long max = (long)Math.ceil(Math.sqrt(UPPER_LIMIT));
        SortedSet<Long> seed = new TreeSet<>();
        for (long i=2; i<=max; i++) {
            seed.add(i);
        }
        squareMap.put(0, seed);
    }

    static int solution(int a, int b) {
        if (b < a) {
            throw new IllegalArgumentException("a should be LE than b");
        }
        if (a < 2) {
            throw new IllegalArgumentException("a should be GE than 2");
        }
        if (b > UPPER_LIMIT) {
            throw new IllegalArgumentException("b should be LE than " + UPPER_LIMIT);
        }
        if (squareMap.isEmpty()) {
            clearAndFillMap();
        }
        if (a == b) {
            for (Map.Entry<Integer, SortedSet<Long>> e : squareMap.entrySet()) {
                if (e.getValue().contains((long)a)) {
                    return e.getKey();
                }
            }
        } else {
            for (Map.Entry<Integer, SortedSet<Long>> e : squareMap.entrySet()) {
                for (long square : e.getValue()) {
                    if (a <= square && square <= b) {
                        return e.getKey();
                    }
                }
            }
        }
        return 0;
    }

    static int solution1(int a, int b) {
        if (a > b) {
            throw new IllegalArgumentException("a should be LE than b");
        }
        if (a < 2) {
            throw new IllegalArgumentException("a should be GE than 2");
        }
        if (b > UPPER_LIMIT) {
            throw new IllegalArgumentException("b should be LE than " + UPPER_LIMIT);
        }

        int result = 0;
        if (a == b) {
            int c = a;
            while ((c = sqrtInt(c)) != -1) {
                result++;
            }
        } else {
            // a < b

        }

        return result;
    }

    static int sqrtInt(int a) {
        double sqrt = Math.sqrt(a);
        int b = (int)Math.round(sqrt);
        return b*b == a ? b : -1;
    }
}

File: exante/src/main/kotlin/exante/Task2.kt
package exante

import java.util.*
import kotlin.math.ceil
import kotlin.math.sqrt

import org.junit.Assert.assertEquals

/**
 * Максимальное кол-во повторяющихся квадратных корней
 */
object Task2 {
    val squareMap = TreeMap<Int, SortedSet<Long>>(Comparator.reverseOrder())
    const val UPPER_LIMIT = 1_000_000_000

    fun solution(a: Int, b: Int): Int {
        if (b < a) {
            throw IllegalArgumentException("a should be LE than b")
        }
        if (a < 2) {
            throw IllegalArgumentException("a should be GE than 2")
        }
        if (b > UPPER_LIMIT) {
            throw IllegalArgumentException("b should be LE than $UPPER_LIMIT")
        }
        if (squareMap.isEmpty()) {
            clearAndFillMap()
        }
        if (a == b) {
            for (e in squareMap.entries) {
                if (e.value.contains(a.toLong())) {
                    return e.key
                }
            }
        } else {
            for (e in squareMap.entries) {
                for (square in e.value) {
                    if (square in a..b) {
                        return e.key
                    }
                }
            }
        }
        return 0
    }

    fun clearAndFillMap() {
        squareMap.clear()
        // fill 0
        fill0()
        var squareRootNumber = 1
        while (true) {
            val previousSquares = squareMap[squareRootNumber-1]!!
            val first = previousSquares.first()
            if (first*first > UPPER_LIMIT) {
                break
            }
            val squares = TreeSet<Long>()
            for (s in previousSquares) {
                val ss = s*s
                if (ss > UPPER_LIMIT) {
                    break
                }
                squares.add(ss)
            }
            squareMap[squareRootNumber] = squares
            squareRootNumber++
        }
    }

    fun fill0() {
        val max = ceil(sqrt(UPPER_LIMIT.toDouble())).toLong()
        val seed = TreeSet<Long>()
        for (i in 2..max) {
            seed.add(i)
        }
        squareMap[0] = seed
    }
}

fun main() {
    assertEquals(4, Task2.solution(2, Task2.UPPER_LIMIT));
    assertEquals(3, Task2.solution(6000, 7000))
    assertEquals(1, Task2.solution(3, 5))
    assertEquals(2, Task2.solution(16, 16))
}